### PR TITLE
SDKv2 to Framework Migration

### PR DESCRIPTION
**Pull request by SBGoods**
- Migrates Time provider and all resources to Framework v0.13.0
- Refactors tests to use `resource.TestCheckResourceAttrSet` helper function
- Adds tests to compare Framework and SDKv2 versions of provider to ensure that there are no differences

### CODE REVIEW

File: internal/provider/provider.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package provider
+
+import (
+	"context"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
```
Comment (bendbennett) [inline internal/provider/provider.go:5]: We may need to sort the imports (see https://github.com/hashicorp/terraform-provider-time/pull/114#discussion_r938640440).

File: internal/provider/provider.go [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+func New() tfsdk.Provider {
+	return &provider{}
+}
+
+var _ tfsdk.Provider = (*provider)(nil)
+
+type provider struct{}
+
+func (p provider) GetSchema(ctx context.Context) (tfsdk.Schema, diag.Diagnostics) {
```
Comment (bendbennett) [inline internal/provider/provider.go:17]: Typically pointer receivers are used for functions defined on types implementing the various interfaces defined in `terraform-plugin-framework` (e.g., `tfsdk.Provider`, `tfsdk.Resource`). There is an example of this in the [online tutorial](https://learn.hashicorp.com/tutorials/terraform/plugin-framework-create?utm_source=WEBSITE&utm_medium=WEB_IO&utm_offer=ARTICLE_PAGE&utm_content=DOCS#explore-the-hashicups-provider).

File: internal/provider/resource_time_offset.go [hunk 3]
```diff
@@ -0,0 +1,455 @@
...
+	offsetYears := types.Int64{Null: true}
+	if idParts[1] != "" {
+		offset, err := strconv.ParseInt(idParts[1], 10, 64)
+		if err != nil {
+			// return diagnostic
+		}
+
+		offsetYears.Value = offset
+		offsetYears.Null = false
+	}
```
Comment (bendbennett) [inline internal/provider/resource_time_offset.go:216]: We can bundle this up into a `func` that we can re-use, perhaps something like:
```go
func offsetToInt64(offsetStr string) (types.Int64, error) {
	offset := types.Int64{Null: true}
	
	if offsetStr != "" {
		offsetInt, err := strconv.ParseInt(offsetStr, 10, 64)
		if err != nil {
			return offset, fmt.Errorf("could not parse offset (%q) as int: %w", offsetStr, err)
		}

		offset.Value = offsetInt
		offset.Null = false
	}
	
	return offset, nil
}
```

If we receive an error, we can generate a diagnostic indicating which offset was problematic and we can also unwrap the error (`unwrappedErr := errors.Unwrap(err)`) in case we need it for logging etc.

File: internal/provider/resource_time_offset.go [hunk 4]
```diff
@@ -0,0 +1,455 @@
...
+
+	var offsetTimestamp time.Time
+
+	if !offsetDays.Null && offsetDays.Value > 0 {
+		offsetTimestamp = timestamp.AddDate(0, 0, int(offsetDays.Value))
+	}
+
+	if !offsetHours.Null && offsetHours.Value != 0 {
+		hours := time.Duration(offsetHours.Value)
+		offsetTimestamp = timestamp.Add(hours * time.Hour)
```
Comment (bendbennett) [inline internal/provider/resource_time_offset.go:293]: Believe we can make the linter happier by using:
```go
		hours := time.Duration(offsetHours.Value) * time.Hour
		offsetTimestamp = timestamp.Add(hours)
```

File: internal/provider/resource_time_offset.go [hunk 5]
```diff
@@ -176,11 +176,110 @@ func (t timeOffsetResourceType) NewResource(ctx context.Context, p tfsdk.Provide
...
 type timeOffsetResource struct {
 }
 
+func (t timeOffsetResource) ModifyPlan(ctx context.Context, req tfsdk.ModifyResourcePlanRequest, resp *tfsdk.ModifyResourcePlanResponse) {
+	// Plan only needs modifying if the resource already exists as the purpose of
+	// the plan modifier is to show updated attribute values on CLI.
+	if req.State.Raw.IsNull() {
+		return
+	}
+
```
Comment (bendbennett) [inline internal/provider/resource_time_offset.go:191]: I think when we were discussing the `ModifyPlan` that I might have skipped over implementing the equivalent for `resourceTimeOffsetConditionExpirationChange()` that's used in the `CustomizeDiff` when running with SDKv2.

Think we'll need something like the following:
```go
	var state, plan timeOffsetModelV0

	diags := req.State.Get(ctx, &state)
	resp.Diagnostics.Append(diags...)
	if resp.Diagnostics.HasError() {
		return
	}

	diags = req.Plan.Get(ctx, &plan)
	resp.Diagnostics.Append(diags...)
	if resp.Diagnostics.HasError() {
		return
	}

	if state.OffsetYears == plan.OffsetYears &&
		state.OffsetMonths == plan.OffsetMonths &&
		state.OffsetDays == plan.OffsetDays &&
		state.OffsetHours == plan.OffsetHours &&
		state.OffsetMinutes == plan.OffsetMinutes &&
		state.OffsetSeconds == plan.OffsetSeconds {
		return
	}
``` 

We'll also need to remove unmarshalling of the plan that currently happens on lines 218-224 as we're now doing that here.

File: internal/provider/resource_time_rotating.go [hunk 6]
```diff
@@ -0,0 +1,710 @@
...
+	Minute          types.Int64  `tfsdk:"minute"`
+	Month           types.Int64  `tfsdk:"month"`
+	RFC3339         types.String `tfsdk:"rfc3339"`
+	Second          types.Int64  `tfsdk:"second"`
+	Unix            types.Int64  `tfsdk:"unix"`
+	Year            types.Int64  `tfsdk:"year"`
+	ID              types.String `tfsdk:"id"`
+}
+
+func setRotationValues(plan *timeRotatingModelV0, timestamp time.Time) (timeRotatingModelV0, error) {
```
Comment (bendbennett) [inline internal/provider/resource_time_rotating.go:480]: The `setRotationValues` could mutate `plan *timeRotatingModelV0` and only return error so that the plan supplied could be used when setting the state.

File: internal/provider/resource_time_rotating.go [hunk 7]
```diff
@@ -0,0 +1,710 @@
...
+		RotationHours:   plan.RotationHours,
+		RotationMinutes: plan.RotationMinutes,
+		RFC3339:         types.String{Value: formattedTimestamp},
+		Unix:            types.Int64{Value: rotationTimestamp.Unix()},
+		ID:              types.String{Value: formattedTimestamp},
+	}, nil
+
+}
+
+func parseTwoPartId(idParts []string, resp *resource.ImportStateResponse) timeRotatingModelV0 {
```
Comment (bendbennett) [inline internal/provider/resource_time_rotating.go:538]: Consider removing `resp *resource.ImportStateResponse` as argument and returning `(timeRotatingModelV0, error)`.

File: internal/provider/resource_time_rotating.go [hunk 8]
```diff
@@ -0,0 +1,710 @@
...
+		RotationDays:    types.Int64{Null: true},
+		RotationHours:   types.Int64{Null: true},
+		RotationMinutes: types.Int64{Null: true},
+		RFC3339:         types.String{Value: formattedTimestamp},
+		Unix:            types.Int64{Value: rotationTimestamp.Unix()},
+		ID:              types.String{Value: formattedTimestamp},
+	}
+}
+
+func parseMultiplePartId(idParts []string, resp *resource.ImportStateResponse) timeRotatingModelV0 {
```
Comment (bendbennett) [inline internal/provider/resource_time_rotating.go:584]: Consider removing `resp *resource.ImportStateResponse` as argument and returning `(timeRotatingModelV0, error)`.

File: internal/provider/resource_time_sleep_test.go [hunk 9]
```diff
@@ -93,27 +170,25 @@ func TestAccTimeSleep_CreateDuration(t *testing.T) {
 				Config: testAccConfigTimeSleepCreateDuration("2ms"),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr(resourceName, "create_duration", "2ms"),
-					testExtractResourceAttr(resourceName, "id", &time2),
-					testCheckAttributeValuesSame(&time1, &time2),
```
Comment (bendbennett) [inline internal/provider/resource_time_sleep_test.go:97]: Do we want to check that the resource has not been recreated?

File: internal/provider/resource_time_sleep_test.go [hunk 10]
```diff
@@ -129,28 +204,27 @@ func TestAccTimeSleep_DestroyDuration(t *testing.T) {
 				Config: testAccConfigTimeSleepDestroyDuration("2ms"),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr(resourceName, "destroy_duration", "2ms"),
-					testExtractResourceAttr(resourceName, "id", &time2),
-					testCheckAttributeValuesSame(&time1, &time2),
```
Comment (bendbennett) [inline internal/provider/resource_time_sleep_test.go:133]: Do we want to check that the resource has not been recreated?

File: internal/provider/resource_time_sleep_test.go [hunk 11]
```diff
@@ -168,14 +242,67 @@ func TestAccTimeSleep_Triggers(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr(resourceName, "triggers.%", "1"),
 					resource.TestCheckResourceAttr(resourceName, "triggers.key1", "value1updated"),
-					testExtractResourceAttr(resourceName, "id", &time2),
-					testCheckAttributeValuesDiffer(&time1, &time2),
```
Comment (bendbennett) [inline internal/provider/resource_time_sleep_test.go:172]: Do we want to check that the resource has been recreated?

File: internal/validators/timevalidator/is_rfc3339_time.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+	return "value must be a string in RFC3339 format"
+}
+
+func (validator isRFC3339TimeValidator) MarkdownDescription(ctx context.Context) string {
+	return validator.Description(ctx)
+}
+
+func (validator isRFC3339TimeValidator) Validate(ctx context.Context, request tfsdk.ValidateAttributeRequest, response *tfsdk.ValidateAttributeResponse) {
+	t := request.AttributeConfig.Type(ctx)
+	if t != types.StringType {
```
Comment (bendbennett) [inline internal/validators/timevalidator/is_rfc3339_time.go:28]: Believe you could use `if !t.Equal(types.StringType)`

**Review by bendbennett — COMMENTED** [review]
Great work!
Looks really good.
Have left a couple of minor comments.

**Review by bendbennett — APPROVED** [review]
LGTM
Great work Selena!

File: internal/provider/resource_time_offset.go [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+				Description: "Number of days to offset the base timestamp. At least one of the 'offset_' arguments must be configured.",
+				Type:        types.Int64Type,
+				Optional:    true,
+				Validators: []tfsdk.AttributeValidator{
+					schemavalidator.AtLeastOneOf(path.MatchRoot("offset_years"),
+						path.MatchRoot("offset_months"),
+						path.MatchRoot("offset_hours"),
+						path.MatchRoot("offset_minutes"),
+						path.MatchRoot("offset_seconds")),
+				},
```
Comment (bflad) [inline internal/provider/resource_time_offset.go:90]: Nit: If we would like to remove declaring this across all the offset attributes, it can be done by adding a declarative configuration validation method, e.g. [`resourcevalidator.AtLeastOneOf`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework-validators@v0.5.0/resourcevalidator#AtLeastOneOf)

```go
import (
  "github.com/hashicorp/terraform-plugin-framework-validators/resourcevalidator"
)

var (
  _ resource.ResourceWithConfigValidators  = (*timeOffsetResource)(nil)
)

func (r timeOffsetResource) ConfigValidators(_ context.Context) []resource.ConfigValidator {
  return []resource.ConfigValidator{
    resourcevalidator.AtLeastOneOf(
      path.MatchRoot("offset_hours"),
      path.MatchRoot("offset_minutes"),
      path.MatchRoot("offset_months"),
      path.MatchRoot("offset_seconds"),
      path.MatchRoot("offset_years"),
    ),
  }
}
```

File: internal/provider/resource_time_offset_test.go [hunk 14]
```diff
@@ -277,6 +459,7 @@ resource "time_offset" "test" {
     %[1]q = %[2]q
   }
   offset_days = 1
+  offset_years = 1
```
Comment (bflad) [inline internal/provider/resource_time_offset_test.go:462]: Was this change intentional?

File: internal/provider/resource_time_rotating.go [hunk 15]
```diff
@@ -0,0 +1,683 @@
...
+				Computed:    true,
+			},
+			"rotation_days": {
+				Description: "Number of days to add to the base timestamp to configure the rotation timestamp. " +
+					"When the current time has passed the rotation timestamp, the resource will trigger recreation. " +
+					"At least one of the 'rotation_' arguments must be configured.",
+				Type:     types.Int64Type,
+				Optional: true,
+				Validators: []tfsdk.AttributeValidator{
+					schemavalidator.AtLeastOneOf(path.MatchRoot("rotation_hours"),
```
Comment (bflad) [inline internal/provider/resource_time_rotating.go:61]: Nit: Similar note about potentially moving the rotation AtLeastOneOf attribute validation to the ConfigValidators method for the resource.

File: internal/provider/resource_time_sleep_test.go [hunk 16]
```diff
@@ -70,13 +148,15 @@ func TestAccTimeSleep_CreateDuration(t *testing.T) {
...
-		CheckDestroy:      nil,
+		ProtoV5ProviderFactories: protoV5ProviderFactories(),
+		CheckDestroy:             nil,
 		Steps: []resource.TestStep{
 			{
 				Config: testAccConfigTimeSleepCreateDuration("1ms"),
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr(resourceName, "create_duration", "1ms"),
+					resource.TestCheckResourceAttrSet(resourceName, "id"),
+					resource.TestCheckResourceAttrSet(resourceName, "id"),
```
Comment (bflad) [inline internal/provider/resource_time_sleep_test.go:159]: Nit:

```suggestion
					resource.TestCheckResourceAttrSet(resourceName, "id"),
```

File: internal/provider/resource_time_sleep_test.go [hunk 17]
```diff
@@ -176,6 +263,59 @@ func TestAccTimeSleep_Triggers(t *testing.T) {
 	})
 }
 
+func TestAccTimeSleep_Upgrade(t *testing.T) {
+	resourceName := "time_sleep.test"
+
+	resource.UnitTest(t, resource.TestCase{
```
Comment (bflad) [inline internal/provider/resource_time_sleep_test.go:269]:
```suggestion
	resource.Test(t, resource.TestCase{
```

File: internal/provider/resource_time_static_test.go [hunk 18]
```diff
@@ -67,8 +66,7 @@ func TestAccTimeStatic_Triggers(t *testing.T) {
 				Check: resource.ComposeTestCheckFunc(
 					resource.TestCheckResourceAttr(resourceName, "triggers.%", "1"),
 					resource.TestCheckResourceAttr(resourceName, "triggers.key1", "value1updated"),
-					testExtractResourceAttr(resourceName, "rfc3339", &time2),
-					testCheckAttributeValuesDiffer(&time1, &time2),
```
Comment (bflad) [inline internal/provider/resource_time_static_test.go:71]: This test was previously verifying whether `triggers` updates ensured the resource was recreated (by expecting a difference in `rfc3339` attribute value). Is this being tested a different way?

File: internal/validators/timevalidator/is_rfc3339_time.go [hunk 19]
```diff
@@ -0,0 +1,61 @@
...
+			"Expected value of type string",
+			t.String(),
+		))
+		return
+	}
+
+	s := request.AttributeConfig.(types.String)
+	if s.Unknown || s.Null {
+		return
+	}
```
Comment (bflad) [inline internal/validators/timevalidator/is_rfc3339_time.go:40]: Nit: Unknown and Null value checking can be on `attr.Value` directly, which can save some minor compute time e.g. at the beginning of this method:

```go
// Only validate the attribute configuration value if it is known.
if req.AttributeConfig.IsNull() || req.AttributeConfig.IsUnknown() {
  return
}
```

File: internal/modifiers/timemodifier/replace_if_outdated.go [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+func ReplaceIfOutdated() tfsdk.AttributePlanModifier {
+	return RequiresReplaceModifier{}
+}
+
+// RequiresReplaceModifier is an AttributePlanModifier that sets RequiresReplace
+// on the attribute.
+type RequiresReplaceModifier struct{}
```
Comment (bflad) [inline internal/modifiers/timemodifier/replace_if_outdated.go:18]: Is this accurate? For reference:

- https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/resource#RequiresReplaceIfFunc
- https://github.com/hashicorp/terraform-plugin-framework/blob/56ca8c935b997cec0a1a9441cc6b5a7cea4fcd2a/resource/plan_modifiers.go#L217

```suggestion
// RequiresReplaceModifier is an AttributePlanModifier that sets RequiresReplace
// on the attribute if the current time is past the stored timestamp.
//
// This custom modifier is necessary because the resource.RequiresReplaceIf
// function uses special logic for Computed attributes which is not applicable
// this use case.
type RequiresReplaceModifier struct{}
```

**Review by bflad — APPROVED** [review]
Looks good to me from a high level (I'm assuming @bendbennett has gone through all the lower level details). 🚀 

Please reach out with any questions from my additional feedback, which is fairly minor in nature.
